### PR TITLE
Add EmbeddingBackward op e2e

### LINKED ISSUE #1267: [Ops] Add support for EmbeddingBW op
Ref https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.embedding_bw.html#ttnn-embedding-bw 

 
- Add corresponding op in TTIR and TTNN dialects. 
- Add runtime support for the op.

### PR DESCRIPTION
**Pull request by rpavlovicTT**
TTNN implements EmbeddingBackward op that computes gradient of the embedding weights w.r.t. input tokens, and gradient of the previous op in backward graph.
This commit adds support for using this op through tt-mlir.

fixes #1267

### CODE REVIEW

File: runtime/lib/ttnn/program.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #include "operations/eltwise/unary/unary.h"
 #include "operations/eltwise/unary/unary_composite.h"
 #include "operations/embedding/embedding.h"
+#include "operations/embedding/embedding_backward.h"
```
Comment (svuckovicTT) [inline runtime/lib/ttnn/program.cpp:20]: Would you mind adding this include (and the `embedding.h` one) to the `tools/ttnn-standalone/ttnn-precompiled.hpp`?

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 2]
```diff
@@ -552,6 +552,26 @@ def TTNN_EmbeddingOp : TTNN_NamedDPSOp<"embedding"> {
...
+def TTNN_EmbeddingBackwardOp : TTNN_NamedDPSOp<"embedding_bw"> {
+    let summary = "Embedding backward op.";
+    let description = [{
+      Embedding backward operation. Generates the gradient of the embedding operation with respect to the input.
+    }];
+
+    let arguments = (ins AnyRankedTensor:$input,
+                         AnyRankedTensor:$weight,
+                         AnyRankedTensor:$in_gradient,
+                         AnyRankedTensor:$output);
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:564]: This doesn't fully reflect the signature of `ttnn::embedding_bw`'s invoke. Add
```
        const std::optional<const DataType> dtype = std::nullopt,
        const std::optional<MemoryConfig>& memory_config = std::nullopt,
```

File: runtime/lib/ttnn/operations/embedding/embedding_backward.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  const ::ttnn::Tensor &inGrad = tensorPool.at(op->in_grad()->global_id());
+  DEBUG_ASSERT(input.is_allocated());
+  DEBUG_ASSERT(weight.is_allocated());
+  DEBUG_ASSERT(inGrad.is_allocated());
+
+  // TTNN implementation of the op gets mem config and dtype out of gradient
+  // tensor argument.
+  ::ttnn::DataType outputDataType = utils::getDataType(op->in_grad());
+  ::ttnn::MemoryConfig outputMemoryConfig =
+      utils::createMemoryConfig(op->in_grad());
```
Comment (svuckovicTT) [inline runtime/lib/ttnn/operations/embedding/embedding_backward.cpp:29]: This should be read from params once they're added in `include/ttmlir/Dialect/TTNN/IR/TTNNOps.td`.

**Review by svuckovicTT — CHANGES_REQUESTED** [review]

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 4]
```diff
@@ -49,6 +51,28 @@ static Value getOrInsertDevice(ConversionPatternRewriter &rewriter,
   return deviceOp.getResult();
 }
 
+static ttnn::ReshapeOp generateReshape(Value input, ArrayRef<int64_t> newShape,
```
[the remark this answers is not in the document]
Comment (rpavlovicTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:54]: I am not sure if that's right way. Right now they are static to this file, as they are not used in any other place. If we put them in utils they will be globally visible. Do you still prefer having them in utils?

btw these 2 functions were just moved within the same file, I did not add them.

[the remark this answers is not in the document]
Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:54]: We can create a utils file for the conversion itself, and put all of these there. [Example](https://github.com/tenstorrent/tt-mlir/pull/1476/files#diff-05f866336c89bd537c3ee19244e211733dcbab69d6654e045ba205568eaccca7). (Note the namespace)

**Review by svuckovicTT — APPROVED** [review]

**Review by kmabeeTT — APPROVED** [review]
Thanks Radenko - rt changes lgtm.

**Review by nsmithtt — APPROVED** [review]

**Pull request merged**